data/bookkeeping/encoding_test.go

```diff
@@ -66,3 +69,145 @@ func TestBlockWithTxnEncoding(t *testing.T) {
 	require.NoError(t, err)
 	require.Equal(t, b, b2)
 }
+
+type typePath []string
+
+func (p typePath) addMapKey() typePath {
+	return append(p, "map_key")
+}
+
+func (p typePath) addValue() typePath {
+	return append(p, "value")
+}
+
+func (p typePath) addField(fieldName string) typePath {
+	return append(p, "field "+fieldName)
+}
+
+func (p typePath) validatePathFrom(t reflect.Type) error {
+	if len(p) == 0 {
+		// path is empty, so it's vacuously valid
+		return nil
+	}
+
+	value := p[0]
+	switch {
+	case value == "map_key":
+		return p[1:].validatePathFrom(t.Key())
+	case value == "value":
+		return p[1:].validatePathFrom(t.Elem())
+	case strings.HasPrefix(value, "field "):
+		fieldName := value[len("field "):]
+		fieldType, ok := t.FieldByName(fieldName)
+		if !ok {
+			return fmt.Errorf("Type '%s' does not have the field '%s'", t.Name(), fieldName)
+		}
+		return p[1:].validatePathFrom(fieldType.Type)
+	default:
+		return fmt.Errorf("Unexpected item in path: %s", value)
+	}
+}
+
+func (p typePath) Equals(other typePath) bool {
+	if len(p) != len(other) {
+		return false
+	}
+	for i := range p {
+		if p[i] != other[i] {
+			return false
+		}
+	}
+	return true
+}
+
+func (p typePath) String() string {
+	return strings.Join(p, "->")
+}
+
+func checkReferencedTypes(seen map[reflect.Type]bool, path typePath, typeStack []reflect.Type, check func(path typePath, stack []reflect.Type) bool) {
+	currentType := typeStack[len(typeStack)-1]
+
+	if _, seenType := seen[currentType]; seenType {
+		return
+	}
+
+	ok := check(path, typeStack)
+	if ok {
+		// if the type is ok, add it to the seen set
+		seen[currentType] = true
+	} else {
+		// if the type is not ok, don't add it to the seen set so we can report more errors if we
+		// see the type again, but we should bail out to avoid checking its children and potentially
+		// getting stuck in an infinite loop
+		return
+	}
+
+	switch currentType.Kind() {
+	case reflect.Map:
+		newPath := path.addMapKey()
+		newStack := append(typeStack, currentType.Key())
+		checkReferencedTypes(seen, newPath, newStack, check)
+		fallthrough
+	case reflect.Array, reflect.Slice, reflect.Ptr:
+		newPath := path.addValue()
+		newStack := append(typeStack, currentType.Elem())
+		checkReferencedTypes(seen, newPath, newStack, check)
+	case reflect.Struct:
+		for i := 0; i < currentType.NumField(); i++ {
+			field := currentType.Field(i)
+			newPath := path.addField(field.Name)
+			newStack := append(typeStack, field.Type)
+			checkReferencedTypes(seen, newPath, newStack, check)
+		}
+	}
+}
+
+func TestBlockFields(t *testing.T) {
+	partitiontest.PartitionTest(t)
+
+	blockType := reflect.TypeOf(Block{})
+
+	// These exceptions are for pre-existing usages of string. Only add to this list if you really need to use string.
+	exceptions := []typePath{
+		typePath{}.addField("BlockHeader").addField("GenesisID"),
```

Comment:
would be nicer to have this as `Payset\[]SignedTxnWithAD\SignedTxn\Txn\AssetConfigTxnFields`, but given that this is a deterministic list, it's quite de-motivating to make the extra effort.